### ISSUE TITLE
MultiLevel Dijkstra (mld) research paper

### ISSUE THREAD

**Issue by AnixPasBesoin** [Project-OSRM/osrm-backend#4797]
**TL;DR.**
I'm looking for is the research paper that the current code implements.
**My issue**
I'm interested in the MLD algorithm implemented in the current version of OSRM.
As far as I've understood, the current implementation allows for independent customization of metrics of the road-network's graph. My first intuition was that this was an implementation of CRP, but after I've read the paper, and gone through some of the OSRM code, I've realized that the **"partition"** step differs from what was proposed in the CRP paper. This made me wonder if CRP was NOT what was implemented here.
So my issue is the following: I'm looking for is the research paper that the current code implements.

**Comment by TimMcCauley** [comment]
@AnixPasBesoin how about this? https://github.com/michaelwegner/CRP

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
@TimMcCauley That's correct - it is an implementation of the following research paper:

Delling, Daniel, et al. "Customizable route planning in road networks." Transportation Science (2015).
https://scholar.google.com/scholar?cluster=5188233396400795462&hl=en

The overall method is called "Customizable route planning," and the "Multi-level Dijkstra" is a part of that method, namely, the query phase.

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
@AnixPasBesoin The partition step is essentially the same - they just implemented it via a different algorithm, that originally does a nested dissection, they adapted it to produce cells.

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
Below is a related issue I've posted and no one ever responded:

https://github.com/Project-OSRM/osrm-backend/issues/4032

**Comment by TimMcCauley** [comment]
@dmitry-akimov-dXgjvQjNiMuVYecJ what does this mean exactly if this were a patent issue?

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
@TimMcCauley I guess, if the algorithm is patented, one cannot use it without Microsoft's permission, and it is probably not a good idea to put it into an open source project.

The question is whether the implementation in OSRM qualifies for the patent. I would say, the method is essentially the same, so it might be the case.

Mapbox is a serious company and might have obtained such permission though.

Therefore, it would be nice if someone from Mapbox clarified this.

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
@TheMarex I see you are responding to other issues. Would you mind saying a couple of words regarding the topic being discussed?

**Comment by AnixPasBesoin** [comment]
After spending some time going through various parts of the current _OSRM_ code, and reading some state-of-the-art papers, here are my conclusions:
 - _MLD_ is (as far as I've understood) the first approach implemented in _OSRM_ that uses _Customizable Metrics_, which makes misleads the reader to think that it's is a pure implementation of _CRP_. 
 - In order to provide a _Multilevel Routing_ for _MLD_, a _Hierarchical Partitioning_ of the road network is needed. I'm assuming that for _Patent Issues_ (as @dmitry-akimov-dXgjvQjNiMuVYecJ mentioned), the _OSRM_ team had to chose a different approach than [_PUNCH_](https://www.microsoft.com/en-us/research/publication/graph-partitioning-with-natural-cuts/), and ultimately, the team has opted for [_Inertial Flow_](http://www.sommer.jp/roadseparator.pdf).
  - At first, I thought that there must be a single paper describing the whole _MLD_ pipeline, but shortly after that, I realized that what I was looking for was the _Partitioning Step_, which is described in the paper I linked above.

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
@AnixPasBesoin I am afraid I did not get what you meant in the part about the MLD and the "Customizable Metrics" - having customizable metrics is kind of the whole point of these algorithms, so what else could it be instead? The MLD is the query phase of the CRP, namely a modification of Dijkstra's algorithm that works on the overlay graph produced by the partitioning step. There is the [bachelor thesis on the topic from KIT](https://i11www.iti.kit.edu/_media/teaching/theses/ba-hamme-13.pdf), which explicitly uses the term "Multilevel Dijsktra" for describing the CRP's query phase. (That bachelor thesis also makes up for a more comprehensive explanation of the method than the research paper, as it usually happens, i.e. if you want to understand a research paper better, find a thesis.)

The method used in OSRM is possibly not a "pure" implementation of the research paper, whatever that might be, but changing the partitioning method does not change the overall method, which is still the CRP to me, at least that is how I would refer to the method used in OSRM. There are other methods, and the OSRM's one is the CRP to me to distinguish it from others.

I must admit I have only skimmed through the CRP research papers, as I had dismissed the method due to the patent issues, and only skimmed through the OSRM source code as well, so I might be missing something, and would appreciate if someone corrects me.

The patent [US20130231862A1](https://patents.google.com/patent/US20130231862A1/en?oq=US20130231862A1) specifically mentions that any partitioning method can be used, and, indeed, it is again not important for the method in its entirety. The OSRM team have implemented InertialFlow, apparently, because it was the simplest method, even though it should produce suboptimal results and is kind of a hack, and proper partitioning methods are quite involved.

I have not gone through the subject in detail, but it still seems to me that the implementation in OSRM might qualify for the patent, and the method is essentially CRP.

I am personally interested in this subject because I have considered CRP as one of the methods I could use in my work, but I skipped it in favor of other methods because of the possible patent issues. I have also considered using OSRM's source code, and now it is unclear what to do with it, as it may be subject to the same patent issues also.

People from Mapbox seem to ignore my questions so far.

**Comment by AnixPasBesoin** [comment]
@dmitry-akimov-dXgjvQjNiMuVYecJ  last time I checked this project, it was using _CH_. Now, in the readme page of the project, _MLD_ is presented as an alternative to _CH_. It is not said that what is meant by _MLD_ is only the query phase of the new approach but is rather presented as the name of the whole, and knowing that _CH_ didn't allow _Customizable Metrics_ it get quite confusing to grasp what changes have been made here since last time.

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
Thankfully, the patent application status has changed to 'abandoned,' so, apparently, the patent is not an issue anymore.

**Comment by dmitry-akimov-dXgjvQjNiMuVYecJ** [comment]
In case anyone interested, it turns out the patent may still be active on this algorithm. But now it seems that we need both a technical specialist and a lawyer to tell. Read more at https://github.com/Project-OSRM/osrm-backend/issues/4032.

**Comment by tenindra** [comment]
Came across this randomly and leaving some additional info. FYI, there is a research paper on a method called ROAD pre-dating CRP by several years: https://dl.acm.org/doi/epdf/10.1145/1458082.1458260 (plus there is a longer follow-up journal paper in TKDE 2012)

The idea of a multi-level Dijkstra algorithm (i.e., hierarchical partitioning with shortcuts between partition border vertices) appears to have originated in the ROAD paper

**Comment by DennisOSRM** [comment]
Thanks for sharing
